Discover and check operator variants

amjames opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76695

Operator variants can now be explicitly specified in the OpInfo kwargs.
When the operator name is not the same as the method/function form this
will allow them to be discovered.

The OpInfo is extended to also accept/discover the inplace operator
variant.

Operator and inplace operator variants are exercised in consistency
tests when the sample does not contain any kwargs.

amjames committed:
Discover and check operator variants

Operator variants can now be explicitly specified in the OpInfo kwargs.
When the operator name is not the same as the method/function form this
will allow them to be discovered.

The OpInfo is extended to also accept/discover the inplace operator
variant.

Operator and inplace operator variants are exercised in consistency
tests when the sample does not contain any kwargs.

[ghstack-poisoned]

mruberry commented on test/test_ops.py:
```diff
@@ -773,16 +773,18 @@ def _case_four_transform(t):
     @_variant_ops(op_db)
     def test_variant_consistency_eager(self, device, dtype, op):
-        # Acquires variants (method variant, inplace variant, aliases)
+        # Acquires variants (method variant, inplace variant, operator variant, inplace_operator variant, aliases)
```
Nice test updates

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -594,6 +594,17 @@ class OpInfo(object):
     # if None, then the OpInfo specifies no method variant
     inplace_variant: Callable = _NOTHING
 
+    # explicitly specifies the operator(module) variant of the operator
```
Nice consistency

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -9824,6 +9850,8 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
                     )),
     BinaryUfuncInfo('bitwise_right_shift',
                     op=torch.bitwise_right_shift,
+                    operator_variant=operator.rshift,
```
Nice specifications

mruberry commented:
Hey @amjames! This looks great but I think you'll need to rebase past some failures on the master branch and then investigate the CI issues -- let me know if you have any questions

amjames committed:
Update on "Discover and check operator variants"


Operator variants can now be explicitly specified in the OpInfo kwargs.
When the operator name is not the same as the method/function form this
will allow them to be discovered.

The OpInfo is extended to also accept/discover the inplace operator
variant.

Operator and inplace operator variants are exercised in consistency
tests when the sample does not contain any kwargs.

[ghstack-poisoned]

amjames commented:
Closing this because I messed up the PR stack and this has been duplicated to #76901
